clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,491 @@
+    if (UO->isIncrementDecrementOp())
+      E = IgnoreCasts(UO->getSubExpr());
+  } else if (const auto *BO = dyn_cast<const BinaryOperator>(S)) {
+    if (BO->isAssignmentOp())
+      E = IgnoreCasts(BO->getLHS());
+  }
+
+  if (E) {
+    if (const auto *D = dyn_cast<DeclRefExpr>(E)) {
+      if (const auto *V = dyn_cast<VarDecl>(D->getDecl())) {
+        if (V->getType()->isCheckedPointerNtArrayType())
+          DefinedVars.insert(V);
+        else {
+
+          // BoundsVars is a mapping from _Nt_array_ptrs to all the variables
+          // used in their bounds exprs. For example:
+
+          // _Nt_array_ptr<char> p : bounds(p + i, i + p + j + 10); 
+          // _Nt_array_ptr<char> q : bounds(i + q, i + p + q + m);
+
+          // EB->BoundsVars: {p: {p, i, j}, q: {i, q, p, m}}
+  
+          for (auto item : EB->BoundsVars) {
+            if (item.second.count(V))
+              DefinedVars.insert(item.first);
+          }
+        }
+      }
+    }
+  }
+
+  for (const auto I : S->children())
+    CollectDefinedVars(I, EB, DefinedVars);
+}
+
+void BoundsAnalysis::ComputeInSets(ElevatedCFGBlock *EB, BlockMapTy BlockMap) {
+  // In[B1] = n Out[B*->B1], where B* are all preds of B1.
+
+  BoundsMapTy Intersections;
+  bool ItersectionEmpty = true;
+
+  for (const CFGBlock *pred : EB->Block->preds()) {
+    if (SkipBlock(pred))
+      continue;
+
+    auto PredEB = BlockMap[pred];
+
+    if (ItersectionEmpty) {
+      Intersections = PredEB->Out[EB->Block];
+      ItersectionEmpty = false;
+    } else
+      Intersections = Intersect(Intersections, PredEB->Out[EB->Block]);
+  }
+
+  EB->In = Intersections;
+}
+
+void BoundsAnalysis::ComputeOutSets(ElevatedCFGBlock *EB,
+                                    BlockMapTy BlockMap,
+                                    WorkListTy &WorkList) {
+  // Out[B1->B2] = (In[B1] - Kill[B1]) u Gen[B1->B2].
+
+  auto Diff = Difference(EB->In, EB->Kill);
+  for (const CFGBlock *succ : EB->Block->succs()) {
+    if (SkipBlock(succ))
+      continue;
+
+    auto OldOut = EB->Out[succ];
+
+    // Here's how we compute (In - Kill) u Gen:
+
+    // 1. If variable p does not exist in (In - Kill), then
+    // (Gen[p] == 0) ==> Out[B1->B2] = {p:1}.
+    // In other words, if p does not exist in (In - Kill) it means that p is
+    // dereferenced for the first time on the incoming edge to this block, like
+    // "if (*p)". So we can initialize the bounds of p to 1. But we may also
+    // run into cases like "if (*(p + 100))". In this case, we cannot
+    // initialize the bounds of p. So additionally we check if Gen[p] == 0.
+
+    // 2. Else if the bounds of p in (In - Kill) == Gen[V] then widen the
+    // bounds of p by 1.
+    // Consider this example:
+    // B1: if (*p) { // In[B1] = {}, Gen[Entry->B1] = {} ==> bounds(p) = 1.
+    // B2:   if (*(p + 1)) { // In[B2] = {p:1}, Gen[B1->B2] = {p:1} ==> bounds(p) = 2.
+    // B3:     if (*(p + 2)) { // In[B2] = {p:2}, Gen[B1->B2] = {p:2} ==> bounds(p) = 3.
+
```

Comment:
At this point, you need to check for each variable with bounds whether the dereference recorded in Gen[succ] widens bounds.  

There are going to be casees to handle:
1. A case where the variable has been widened, i.e. has an entry in Diff.
2. A case the variable has not been widened, i.e. does not have an entry diff.